### PR TITLE
PhpdocAlignFixer - add align option

### LINKED ISSUE #3450: Disable phpdoc_align?
I there any way to **disable** vertical align for phpDocs?

Before and after:
```php
    /**
     * @param            $jobClass
     * @group             functional
     * @dataProvider dataJobCreation
     */
```
What I expect:
```php
    /**
     * @param $jobClass
     * @group functional
     * @dataProvider dataJobCreation
     */
```
Config field: `'phpdoc_align' => false` or `'phpdoc_align' => ['tags' => []]` (some other field?).
Command: `php-cs-fixer fix --config=.php_cs foo/Bar.php`.
CS Fixer `2.10.0`.
PHP `7.1.13`.

### PR DESCRIPTION
**Pull request by zored**
Closes #3450

### CODE REVIEW

File: src/Fixer/Phpdoc/PhpdocAlignFixer.php [hunk 1]
```diff
@@ -85,6 +86,7 @@ public function configure(array $configuration = null)
 
         $this->regex = '/^'.$indent.' \* @(?:'.implode('|', $types).')'.$desc.'\s*$/u';
         $this->regexCommentLine = '/^'.$indent.' \*(?! @)(?:\s+(?P<desc>\V+))(?<!\*\/)$/u';
+        $this->collapse = (bool) $this->configuration['collapse'];
```
Comment (keradus) [inline src/Fixer/Phpdoc/PhpdocAlignFixer.php]: casting is pointless, for non-boolean values `FixerConfigurationResolver` will raise exception

Comment (zored) [reply src/Fixer/Phpdoc/PhpdocAlignFixer.php]: fixed, cleaned CS, reran builds)

**Comment by zored** [comment]
Found little bugs for these inputs in my project.

- Inital spaces removed from single-line descriptions:
```
    /**
     * @param string | \Closure          $value If closure - it must have the folloting signature:
     * function ($item) {
     *  return value;
     * }
     * @param callback|null              $valueCallback
     * @return array
     */
```
- Extra space added:
```
    /**
     * @param         $id
     * @param         $parentId
     * @param int     $websiteId
     * @param         $position
     * @param int[][] $siblings
     *
     * @dataProvider dataMoveCategorySameParent
     */
```

File: tests/Fixer/Phpdoc/PhpdocAlignFixerTest.php [hunk 2]
```diff
@@ -79,6 +79,68 @@ public function testFixCollapse()
...
+    }
+
+    public function testFixPartiallyUntypedCollapse()
+    {
+        $this->fixer->configure(['tags' => ['param'], 'collapse' => true]);
+
+        $expected = <<<'EOF'
+<?php
+    /**
+     * @param $id
```
Comment (zored) [inline tests/Fixer/Phpdoc/PhpdocAlignFixerTest.php]: fixed only bug I found locally for my new flag: partially untyped annotations came with extra spaces

**Comment by SpacePossum** [comment]
Hi,
With the new option; would fixing like the following be possible?

```php
<?php

class SomeTest
{
    /**
     * @param int $a
     * @param string $b
     * 
     * @dataProvider           dataJobCreation
     */
    public function testSomething($a, $b)
    {
    }
}
```

```php
<?php

class SomeTest
{
    /**
     * @param int    $a
     * @param string $b
     * 
     * @dataProvider dataJobCreation
     */
    public function testSomething($a, $b)
    {
    }
}
```

**Comment by zored** [comment]
@SpacePossum hi! What you show is the default behavior without `'collapse' => true`. New flag forbids vertical align for shorter diffs in pull-requests.

Maybe I misunderstood the question and I need to clarify my solution with additional tests. 😃

**Comment by zored** [comment]
Added `dataProvider` support.

**Comment by zored** [comment]
Any other comments? :)

File: src/Fixer/Phpdoc/PhpdocAlignFixer.php [hunk 3]
```diff
@@ -165,7 +168,13 @@ protected function createConfigurationDefinition()
             ])
         ;
 
-        return new FixerConfigurationResolver([$tags->getOption()]);
+        $collapse = (new FixerOptionBuilder('collapse', 'Collapse spaces (disables vertical align).'));
+        $collapse
+            ->setAllowedTypes(['bool'])
+            ->setDefault($this->collapse)
```
Comment (keradus) [inline src/Fixer/Phpdoc/PhpdocAlignFixer.php]: add new `CodeSample` that uses this option

File: src/Fixer/Phpdoc/PhpdocAlignFixer.php [hunk 4]
```diff
@@ -44,6 +45,7 @@
         'type',
         'var',
         'method',
+        'dataProvider',
```
Comment (keradus) [inline src/Fixer/Phpdoc/PhpdocAlignFixer.php]: this one line and tests for it shall be raised against `2.2` branch

Comment (zored) [reply src/Fixer/Phpdoc/PhpdocAlignFixer.php]: Removed.

File: README.rst [hunk 5]
```diff
@@ -1171,6 +1171,8 @@ Choose from the list of available rules:
 
   Configuration options:
 
+  - ``collapse`` (``bool``): collapse spaces (disables vertical align); defaults to
```
[the remark this answers is not in the document]
Comment (keradus) [reply README.rst]: while this point is valid, let us postpone the fixer rename in sake to making PR mergeable sooner.

could you please review all the rest ?

File: src/Fixer/Phpdoc/PhpdocAlignFixer.php [hunk 6]
```diff
@@ -35,6 +35,7 @@
 {
     private $regex;
     private $regexCommentLine;
+    private $collapse = false;
```
Comment (kubawerlos) [inline src/Fixer/Phpdoc/PhpdocAlignFixer.php]: The default value is useless as it always will be overridden in `configure`.

File: README.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (zored) [reply README.rst]: @julienfalque, @keradus, we can do easy fix: like CSS `text-align: left` and `justify` values we can use `'align' => 'left'` and `'align' => 'vertical'` by default.
What do you think?

File: src/Fixer/Phpdoc/PhpdocAlignFixer.php (hunk 6 above)
Comment (zored) [reply src/Fixer/Phpdoc/PhpdocAlignFixer.php]: We can refactor this config a bit. You are welcome [to the discussion](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3453#discussion_r169171531). 😃

Comment (kubawerlos) [reply src/Fixer/Phpdoc/PhpdocAlignFixer.php]: No much to refactor, it's simply about removing ` = false` part - real setting of default value is done `createConfigurationDefinition` method anyway.

File: README.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (julienfalque) [reply README.rst]: :+1:  for `align => vertical|left`.

**Comment by zored** [comment]
Merge, please 😄

**Comment by SpacePossum** [comment]
Rebased, conflict resolved, CS fixed, squashed and forced pushed.
@zored please feel free to reclaim the authorship of the commit.

**Comment by zored** [comment]
@SpacePossum thanks. Let commit be yours :)

**Comment by keradus** [comment]
Thank you @zored.

**Pull request merged**